CT-279 Use netty-based AsyncHttpClient in AddEventsClient

oliverhsu77 opened the pull request:
Switch for Apache HTTP client to Netty-based `AsyncHttpClient` for improved performance and throughput.

Kami commented on src/main/java/com/scalyr/integrations/kafka/AddEventsClient.java:
```diff
@@ -215,47 +221,53 @@ private AddEventsResponse addEventsWithRetry(byte[] addEventsPayload, long uncom
+      // error - retry if not timed out and retriable, otherwise, complete with failure
+      log.warn("addEvents failed with {}", addEventsResponse);
+      if (remainingMs(startTimeMs) > retryDelayMs && addEventsResponse.isRetriable()) {
+        sleep.accept(retryDelayMs);
```
I'm not that familiar with Java async patterns, but I assume ``Uninterruptibles.sleepUninterruptibly`` plays along nicely with the async code and doesn't block the whole event loop?

oliverhsu77 replied:
` Uninterruptibles.sleepUninterruptibly` does block.  I changed this to use a `ScheduledExecutorService`.  See response on line 79 for more details.

Kami commented on src/main/java/com/scalyr/integrations/kafka/AddEventsClient.java:
```diff
@@ -215,47 +221,53 @@ private AddEventsResponse addEventsWithRetry(byte[] addEventsPayload, long uncom
-    long delayTimeMs = initialBackoffDelayMs;
-    httpPost.setEntity(new ByteArrayEntity(addEventsPayload));
+    // Retries timed out
+    if (remainingMs(startTimeMs) == 0) {
```
(just thhinking out loud) - I was wondering if ``remainingMs()`` can ever return a negative value, but I see it's guarded with max clause which will return 0 in such cases.

Kami commented on src/main/java/com/scalyr/integrations/kafka/AddEventsClient.java:
```diff
@@ -74,10 +73,10 @@
-  private final CloseableHttpClient client = HttpClients.createDefault();
+  private static final AsyncHttpClient asyncHttpClient = asyncHttpClient();
   private final ObjectMapper objectMapper = new ObjectMapper();
-  private final ExecutorService senderThread = Executors.newSingleThreadExecutor();
```
This question is more due to my limited understanding of async patterns in Java - does it still make sense to use two threads here when using async http client?

In other languages, it depends, but in most cases when you are using async code you don't want to use multiple threads in such scenario, but a single one (this way you actually utilize resources efficiently and you can do that since async code won't block other async code during blocking / waiting operations).

I imagine not using two threads here may also require more code to be refactored and rely on some kind of other "async native" primitive for sleep, but dunno.

oliverhsu77 replied:
Addressing both the `Uninterruptibles.sleepUninterruptibly(timeMs, TimeUnit.MILLISECONDS)` and this comment together since they are related.

For `ListenableFuture.addListener(Runnable listener, @Nullable Executor exec)`, if `exec` is null, the callback is executed on the Netty thread, and we should not block or tie up this thread.
https://github.com/AsyncHttpClient/async-http-client#setting-callbacks-on-the-listenablefuture

Since the retry has a delay, we don't want to tie up the Netty thread when `Uninterruptibles.sleepUninterruptibly`sleeps on retry.  That's where this thread pool comes in.  It is possible to use a `ScheduledExecutorService` instead.  There are some advantages to this when we have multiple tasks running in the worker node and they can use the same single `ScheduledExecutorService` thread for the delay instead of 1 thread per task.  This would require more code refactoring to go this route.  It is also possible o use the RxJava library to do an "async native sleep".

I'll look into this to see if there's something we can do without a lot of code refactoring.

Kami reviewed (approved):
Overall, changes look good to me, but my understanding of async primitives and patterns in Java is limited.

czerwingithub commented on src/test/java/com/scalyr/integrations/kafka/ScalyrSinkTaskTest.java:
```diff
@@ -507,6 +508,7 @@ public void testFlush() {
   /**
    * Verify stop doesn't throw any exceptions
    */
+  @Ignore("Test closes AddEventsClient AsyncHttpClient, which causes subsequent tests to fail")
```
Ok, are there problems with us not being able to call stop?

oliverhsu77 replied:
I thought about this some more.  The `AsyncHttpClient` life cycle should match the lifecycle of the connector.  When the task stops, we should not close the `AsyncHttpClient`.  I changed this to have the `AsyncHttpClient` match the lifecycle for the connector.

czerwingithub commented on src/main/java/com/scalyr/integrations/kafka/AddEventsClient.java:
```diff
@@ -74,10 +73,10 @@
-  private final ExecutorService senderThread = Executors.newSingleThreadExecutor();
-  private final HttpPost httpPost;
+  private final ExecutorService retryThread = Executors.newSingleThreadExecutor();
+  private final BoundRequestBuilder requestBuilder;
```
Can you add documentation for this variable?  The bound thing raises an eyebrow for me.

czerwingithub reviewed (changes requested):
Just a few comment.  Tomaz raised a few good points that I'll be interested on as well.

oliverhsu77 commented:
Ready for another look.  Made the following changes:
1. Addressed issues with `AsyncHttpClient` life cycle.  Life cycle should be based on the connector life cycle and not the task life cycle.
2. Change retry delay to use `ScheduledExecutorService` instead of `Uninterruptibles.sleepUninterruptibly`, so thread will not be blocked during the sleep.  Also changed the http request async listener to run on the IO thread, since this listener no longer contains blocking / long running operations.

Kami reviewed (approved):
Thanks for the additional clarification.

Overall the change looks good to me, but I wonder how we plan to release this - I assume we will want to perform at least some very basic benchmarking / load test before publishing new release to maven?

I know that was already kinda on your TODO, just wanted to make sure we are all on the same page.

oliverhsu77 commented:
Yes, I think it would be good to perform some benchmarks before making a Github release.
